src/transmem.cpp

```diff
@@ -4,39 +4,286 @@
+    StampedAndRatedTransformation resultingTransformation;
+    resultingTransformation.time = tstamp;
+    calculateTransformation(p, resultingTransformation);
+
+    return resultingTransformation;
+}
+
+StampedAndRatedTransformation TransMem::getBestLink(const FrameID &srcFrame, const FrameID &destFrame) {
+
+    if(srcFrame == destFrame)
+        throw std::invalid_argument("Not allowed to insert a link with srcFrame == destFrame.");
+
+    bool recalculation = true;
+    std::string linkID = srcFrame+destFrame;
+
+    std::lock_guard<std::recursive_mutex> guard(lock);
+
+    // Check if link is already cached.
+    auto itr2CachedBestLink = cachedBestLinks.find(linkID);
+    if(itr2CachedBestLink != cachedBestLinks.end()){
+
+    Timestamp tstampBestLinkCalc = (*itr2CachedBestLink).second.first;
+    Path pathBestLink = (*itr2CachedBestLink).second.second;
+
+    // The link is just recalculated if every link was updated in the mean time.
+       for(Link& l : pathBestLink.links)
+           recalculation = recalculation && ( tstampBestLinkCalc < l.lastTimeUpdated);
+   }
+
+    // Do the recalculation if either the link was not cached before or every link along
+    // the path was updated since the link was added to the cache.
+    if(recalculation){
+
+        StampedAndRatedTransformation stT;
+        Path path{srcFrame, destFrame};
+
+        if(!bestLink(stT, path))
+               throw NoSuchLinkFoundException(srcFrame, destFrame);
+
+        // Remove old entries.
+        cachedBestLinks.erase(linkID);
+        cachedBestTransformations.erase(linkID);
+
+        cachedBestLinks.insert({linkID,{std::chrono::high_resolution_clock::now(), path}});
+        cachedBestTransformations.insert({linkID, stT});
+
+        return stT;
+    }
+    else{
+        return cachedBestTransformations.at(linkID);
+    }
+}
+
+// Public debug functions
+
+void TransMem::dumpAsJSON() const {
+
+    QString path = "";
+    QJsonObject transmemObject;
+
+    std::lock_guard<std::recursive_mutex> guard(lock);
+
+    writeJSON(transmemObject);
+
+    dumpJSONfile(path, transmemObject, OutputType::TRANSMEM);
+
+    return;
+}
+
+void TransMem::dumpAsGraphML() const {
+
+   GraphMLWriter writer;
+   QString path = "";
+
+   std::lock_guard<std::recursive_mutex> guard(lock);
+   writer.write(path, *this);
+
+}
+
+// Protected functions
+
+bool TransMem::bestLink(StampedAndRatedTransformation &stT, Path &p) const {
+
+    // Asume the lock is already aquired.
+
+    // Search for shortest path between source frame and  destination frame.
+    if(!shortestPath(p))
+        return false;
+
+    // Evaluate best point in time.
+    calculateBestPointInTime(p, stT.time);
+
+    // Calculate transformation along path.
+    calculateTransformation(p, stT);
+
+    return true;
+}
+
+void TransMem::calculateTransformation(const Path &path, StampedAndRatedTransformation &resultT) const {
+
+    // Asume the lock is already aquired.
```

Comment:
This is something that should also be publicly documented. The TransMem class should document the locking behaviour (all public function acquire lock?).